### PR TITLE
Adding HAPCut functionality to `astroquery.mast.Cutouts` tool

### PR DESCRIPTION
**Pull request by jaymedina**
This PR will expand `astroquery.mast`'s cutout functionality to support making Hubble Advanced Products (HAP) cutouts via MAST's [`HAPCut` API](https://mast.stsci.edu/hapcut/getting_started.html). 

! **Timeline: ASAB would like to see this merged before the end of Q1. Ideally before Dec 21st if possible.**

Changes include: 
- [x] New `HapcutClass` in `cutouts.py`
- [x] Unit testing for `HapcutClass` 
- [x] Updated documentation for `HapcutClass` functionality 
- [x] Updated release notes

## Example `download_cutouts` call with sample cutout shown: 

image

#### (default cutout size is 5 x 5 pixels): 

image

### CODE REVIEW

**Comment by jaymedina** [comment]
This question came up during one of our stand-ups: When will the next release of `astroquery` be?

File: docs/mast/mast.rst
Comment (bsipocz) [inline docs/mast/mast.rst:1137]:
```suggestion
an angular quantity) and returns the cutout FITS file(s) as a list of `~astropy.io.fits.HDUList` objects.
```

File: docs/mast/mast.rst
Comment (bsipocz) [inline docs/mast/mast.rst:1123]: side note: the doctest are failing in an earlier example, so these has not been run during my testing

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:703]: what's the purpose of this line?

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:657]: There seem to be a lot of duplicates in this method shared amongst the different cutout classes, I wonder maybe it's time to consolidate them into a base class, so any future maintenance becomes easier

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:721]: Don't use prints, us the logger or a verbose keyword instead. This should be changed in the other classes, too.

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:772]: extreme nitpick: fyi: we allow longer than 80char lines, no need to cut this into 3

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:781]: I think we should distinguish between an error in the response as opposed to a null response. I agree that it should not be an exception if there are no results, but if the response contains a genuine error message, we should raise that rather than warn about it.

File: astroquery/mast/cutouts.py
Comment (bsipocz) [inline astroquery/mast/cutouts.py:780]: Is there any reasons why response.json() rather than e.g. response.text is used?

File: astroquery/mast/tests/test_mast_remote.py
Comment (bsipocz) [inline astroquery/mast/tests/test_mast_remote.py:1101]: nitpick: ideally tests like these can be parametrized, so they would be independent and would not require repeated code

File: astroquery/mast/tests/test_mast_remote.py
Comment (bsipocz) [inline astroquery/mast/tests/test_mast_remote.py:1107]: could we check the shape for the other size inputs, too?

**Review by bsipocz — APPROVED** [review]
I have some big picture comments, which certainly points beyond this current PR.:

- I wonder at what point will it make sense to make a generic mast.Cutout class, that takes the coordinate and returns all available cutouts, or maybe also takes a missing name? Even if it's not exposed to the public API, I think we reached a point where having mast cutout base class would be beneficial as some of the methods share significant amount of code (e.g. enhancements and bugfixes, or any maintenance will be easier with a baseclass).


Otherwise I have minor comments, so would only like to see a few small things to be fixed.

File: docs/mast/mast.rst
Comment (jaymedina) [reply docs/mast/mast.rst:1148]: The internal database that these calls access changes pretty regularly so I'll have to ignore the doctest.

Comment (bsipocz) [reply docs/mast/mast.rst:1148]: but you're not testing the pixel data here, and I presume the dimension and format of the cutout should stay the same, would it not?

Comment (bsipocz) [reply docs/mast/mast.rst:1148]: OK, thanks for showing the details. Agree on ignoring the output then.

File: astroquery/mast/cutouts.py
Comment (jaymedina) [reply astroquery/mast/cutouts.py:780]: I don't think so, but I would defer to @ceb8 for that one. Using `response.text` requires you to index with an integer for the `msg` keyword value, so I'm guessing it's probably safer to index `response.json()` explicitly for `msg` rather than an implicit integer index of the `response.text` output.

Comment (bsipocz) [reply astroquery/mast/cutouts.py:778]: OK, indeed makes sense. Thanks.

Maybe, can we add another assert, or even better add a `match=` in the `pytest.warn` line, so check we indeed parse out the expected message.

Comment (bsipocz) [reply astroquery/mast/cutouts.py:778]: nope, keep the json, but add the match parameter to pytest.warns() for the tests, so it will only pass if the warning message also matching not just the type

File: astroquery/mast/tests/test_mast_remote.py
Comment (jaymedina) [reply astroquery/mast/tests/test_mast_remote.py:1107]: Thanks you just helped me catch a bug. Updated.

File: astroquery/mast/cutouts.py
Comment (jaymedina) [reply astroquery/mast/cutouts.py:780]: Will do

File: astroquery/mast/cutouts.py
Comment (jaymedina) [reply astroquery/mast/cutouts.py]: I made the distinction by way of an `if`/`else` statement. Let me know if this is what you were thinking of.

**Comment by jaymedina** [comment]
> I wonder at what point will it make sense to make a generic mast.Cutout class, that takes the coordinate and returns all available cutouts, or maybe also takes a missing name? Even if it's not exposed to the public API, I think we reached a point where having mast cutout base class would be beneficial as some of the methods share significant amount of code (e.g. enhancements and bugfixes, or any maintenance will be easier with a baseclass).

This has actually been a topic of discussion recently, and we do have plans to consolidate all the cutouts classes into a single mission-agnostic cutout tool. I'm not sure what the timeline is will be on this, though. We have a few more things we want to update `astroquery.mast` with, including updating `TesscutClass` with support for a high level science product called TICA, as well as integrating `Cutouts` to the cloud, so consolidating all the classes into one will likely happen after that.

File: astroquery/mast/tests/test_mast_remote.py
Comment (jaymedina) [reply astroquery/mast/tests/test_mast_remote.py:1101]: Agreed, I think parametrization of some of these tests will be straight forward after consolidating all the classes into 1. One problem I see us running into is that some of the example inputs we use (mainly for the `coordinates` arg) only work in one tool and not the other, e.g. and RA/Dec of `102.7, 70.50` will work for TESSCut and not HAPCut. So I'll need to find a set of parameters that will work for all 3. But that's a bridge we can cross when we get to it.

**Comment by bsipocz** [comment]
> This has actually been a topic of discussion recently ...

Sounds all great. It certainly points beyond this current PR, and any timeline will work that works for you.
Also re: cloud - as part of fornax (so mast is in the know), we're factoring out the cloud download logic from the mast module and upstream it into pyvo for a more generic audience, so it would be useful to communicate any plans for enhancements if it affects the bottom infrastructure.

File: astroquery/mast/tests/test_mast_remote.py
Comment (bsipocz) [reply astroquery/mast/tests/test_mast_remote.py:1101]: oh, I didn't mean to parametrize between classes, but for `size` within each of these tests, with a `pytest.mark.parametrize` (as practically that's the only thing changing within the test)

**Comment by bsipocz** [comment]
Could you squash out some of the later debugging commits? I plan to come back and go through it again in the evening, but don't think there is any more things to address for merge.

**Comment by jaymedina** [comment]
Yeah, I'm waiting for a science review on this from one group, once I hear back from them I'll squash this down so it's ready for merge.

**Comment by jaymedina** [comment]
Squashed! @bsipocz

**Comment by bsipocz** [comment]
Thanks! No need to squash to one, it's really more about keeping the incremental fixes at the later parts into more logical chunks. 
Either way, this is now good to go and I'll do it later today, but first will reuse a PR for an experiment for the gh command line client.

**Comment by bsipocz** [comment]
Thank you @jaymedina!

**Pull request merged**
